### PR TITLE
Websockets

### PR DESCRIPTION
**Pull request by amanwriter**
Added support for Websockets. Host, tcp_bus now store services in a dict to reduce replicated code.

### CODE REVIEW

File: examples/wschat.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+                    for m in self.msgs:
+                        messages += m + "<br>"
+                    wsk.send_str(messages)
+                else:
+                    self.msgs.append(msg.data)
+        return wsk
+
+if __name__ == '__main__':
+    wss = ChatWSService('0.0.0.0', 4501)
+    Host.registry_host = '127.0.0.1'
```
Comment (kashifrazzaqui) [inline examples/wschat.py]: Use Host.configure

File: examples/wschat.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    def __init__(self, ip, port):
+        super(ChatWSService, self).__init__("ChatService", "1", ip, port)
+        self.msgs = []
+
+    @ws(path='/')
+    def chat(self, request):
+        wsk = WebSocketResponse()
+        wsk.start(request)
+        while True:
```
Comment (kashifrazzaqui) [inline examples/wschat.py]: This looks awfully like the HTTPService but that is request-response based which doesn't have stream semantics like **wsk.receive()**. This should be more like TCPService where the TCP socket is not exposed.

Comment (amanwriter) [inline examples/wschat.py]: Is this design okay:

``` python
@ws(path='/')
def chat(msg: str) -> str:
   #process message and generate response
   response = process_msg(msg)
   return response
```

This should hide the WebSocketResponse and the while loop from the service and it can be moved to the decorator.

**Comment by nerandell** [comment]
@kashifrazzaqui Can you take a look now. We have hidden the socket from the user of the service now.

File: examples/wschat.py (hunk 2 above)
Comment (kashifrazzaqui) [inline examples/wschat.py]: This looks nice, but I would like to see the actual code.

**Comment by nerandell** [comment]
Fixes #55

File: examples/wschat.html [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    <button onclick="send()">Send</button>
+
+</center>
+
+<script>
+
+var ws = new WebSocket("ws://127.0.0.1:4501");
+
+ws.onopen = function(){
+var timertoggle = setInterval(function(){ws.send('updater');}, 1000);
```
Comment (nerandell) [inline examples/wschat.html]: PEP8

File: dump.rdb [hunk 4]
```diff
@@ -0,0 +1 @@
+REDIS0006�ܳC�Z��V
```
Comment (nerandell) [inline dump.rdb]: Why is this file here? use `git rm --cached`

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,3 @@
+FROM docker2.1mg.com:5000/1mg/vyked
+ADD . /opt/vyked
+RUN cd /opt/ && pip3 install -e vyked
```
Comment (nerandell) [inline Dockerfile]: Is this file required?

File: vyked/host.py [hunk 6]
```diff
@@ -59,16 +58,19 @@ def configure(cls, name, registry_host: str = "0.0.0.0", registry_port: int = 45
...
-            cls._tcp_service = service
-        else:
+        invalid_service = True
+        _service_classes = {'_tcp_service': TCPService, '_http_service': HTTPService, '_ws_service': WSService}
+        for key, value in _service_classes.items():
+            if isinstance(service, value):
+                cls._services[key] = service
+                invalid_service = False
+                break
+        if invalid_service:
```
Comment (nerandell) [inline vyked/host.py]: for ``for .. else`` from python.

File: vyked/services.py [hunk 7]
```diff
@@ -321,6 +341,10 @@ def delete(path=None, required_params=None, timeout=API_TIMEOUT):
     return get_decorated_fun('delete', path, required_params, timeout)
 
 
+def ws(path=None, required_params=None, timeout=None):
```
Comment (nerandell) [inline vyked/services.py]: Rename it from `ws` to `websocket`. Makes more readable for the end-user

File: vyked/services.py [hunk 8]
```diff
@@ -579,3 +603,28 @@ def _send_http_request(self, app_name, method, entity, params):
         response = yield from self._http_bus.send_http_request(app_name, self.name, self.version, method, entity,
                                                                params)
         return response
+
+
+class WSService(_ServiceHost, metaclass=OrderedClassMembers):
```
Comment (nerandell) [inline vyked/services.py]: Lots of duplication between `HTTPService` and `WSService`. Just extract out all the common code `HTTPService` and `WSService` can extend these classes.

File: vyked/registry.py [hunk 9]
```diff
@@ -296,12 +296,12 @@ def _connect_to_service(self, host, port, node_id, service_type):
                 coroutine = self._loop.create_connection(partial(get_vyked_protocol, self), host, port)
                 future = asyncio.async(coroutine)
                 future.add_done_callback(partial(self._handle_service_connection, node_id, host, port))
-        elif service_type == 'http':
+        elif service_type == 'ws' or service_type == 'http':
```
Comment (nerandell) [inline vyked/registry.py]: `elif service_type in ('ws', 'http')`. Looks more short and concise but that's just my opinion.

File: vyked/host.py [hunk 10]
```diff
@@ -164,16 +179,16 @@ def _set_host_id(cls):
...
-        if tcp_server:
-            cls._logger.info('Serving TCP on {}'.format(tcp_server.sockets[0].getsockname()))
-        if http_server:
-            cls._logger.info('Serving HTTP on {}'.format(http_server.sockets[0].getsockname()))
+            for service in cls._services.values():
+                if service:
+                    asyncio.get_event_loop().run_until_complete(service.tcp_bus.connect())
+        for key, server in server_dict.items():
+            if server:
+                cls._logger.info('Serving '+key+' on {}'.format(server.sockets[0].getsockname()))
```
Comment (nerandell) [inline vyked/host.py]: PEP 8 in the line.

File: examples/wschat.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    def __init__(self, ip, port):
+        super(ChatWSService, self).__init__("ChatService", "1", ip, port)
+        self.msgs = []
+
+    @ws(path='/')
+    def chat(self, msg=None):
+        if 'updater' in msg.data:
+            messages = ""
+            for m in self.msgs:
+                messages += m + "<br>"
```
Comment (nerandell) [inline examples/wschat.py]: `messages = "<br>".join(self.msgs)`
`messages+="<br>"`
`for .. loop` in this case is not very pythonic.
